apisix/plugins/request-validation.lua

```diff
@@ -0,0 +1,84 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core          = require("apisix.core")
+local plugin_name   = "request-validation"
+local json_decode   = require("cjson").decode
+local ngx           = ngx
+
+
+local schema = {
+    type = "object",
+    properties = {
+        body_schema = {type = "object"},
+        header_schema = {type = "object"}
+    },
+    oneOf = {
+        {required = {"body_schema"}},
+        {required = {"header_schema"}}
+    }
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 2800,
+    type = 'validation',
+    name = plugin_name,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.rewrite(conf)
+    if conf.body_schema.properties.header_schema then
+        local headers = ngx.req.get_headers()
+        local ok, err = core.schema.check(conf.header_schema, headers)
+        if not ok then
+            core.log.error("req schema validation failed", err)
+            core.response.exit(400, err)
+        end
+    end
+
+    if not conf.body_schema.properties.body_schema then
+        ngx.req.read_body()
+        local body = ngx.req.get_body_data()
+        if not body then
+            body = ngx.req.get_body_file()
```

Comment:
the current way is wrong.

`file_name = ngx.req.get_body_file()`

https://github.com/openresty/lua-nginx-module#ngxreqget_body_file

Reply:
Shall I not read the body file, as its anyways going to be big? The following is the sample code I got to read the body file. And it seems, reading file will block the even loop of nginx.
```lua
local filename = ngx.req.get_body_file()
if not filename then
    return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
end
-- WARNING
-- don't use this code in production, file I/O is blocking,
-- you are going to block nginx event loop at this point!
local fd = io.open(filename, 'rb')
if not fd then
    return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
end
body = fd:read('*a')
```

Reply:
if we use `jsonschema` to check the request body, we have to do it in this way.

Reply:
If would be nice if a json formatted error is returned, Rather than the generic page